components/scream/src/share/util/scream_kokkos_utils.hpp

```diff
@@ -68,6 +72,123 @@ reshape (Kokkos::View<DataTypeIn,Props...> view_in,
   return view_out;
 }
 
+/*
+ * ExeSpaceUtils is essentially a TeamPolicy factory. TeamPolicy objects
+ * are what kokkos uses to define a thread layout (num teams, threads/team)
+ * for a parallel kernel. On non-GPU archictures, we will generally have
+ * thread teams of 1.
+ */
+template <typename ExeSpace = Kokkos::DefaultExecutionSpace>
+struct ExeSpaceUtils {
+  using TeamPolicy = Kokkos::TeamPolicy<ExeSpace>;
+
+  static TeamPolicy get_default_team_policy (Int ni, Int nk) {
+#ifdef MIMIC_GPU
+    const int max_threads = ExeSpace::concurrency();
+    const int team_size = max_threads < 7 ? max_threads : 7;
+    return TeamPolicy(ni, team_size);
+#else
+    return TeamPolicy(ni, 1);
+#endif
+  }
+
+  static TeamPolicy get_team_policy_force_team_size (Int ni, Int team_size) {
+    return TeamPolicy(ni, team_size);
+  }
+};
+
+/*
+ * Specialization of above for Cuda execution space. Many GPU architectures can
+ * support a great number of threads, so we'll need to expose additional
+ * parallelism by having many threads per team.  This is due to having more
+ * threads than the main kernel loop has indices.
+ */
+#ifdef KOKKOS_ENABLE_CUDA
+template <>
+struct ExeSpaceUtils<Kokkos::Cuda> {
+  using TeamPolicy = Kokkos::TeamPolicy<Kokkos::Cuda>;
+
+  static TeamPolicy get_default_team_policy (Int ni, Int nk) {
+    return TeamPolicy(ni, std::min(128, 32*((nk + 31)/32)));
+  }
+
+  static TeamPolicy get_team_policy_force_team_size (Int ni, Int team_size) {
+    return TeamPolicy(ni, team_size);
+  }
+};
+#endif
+
+/*
+ * TeamUtils contains utilities for getting concurrency info for
+ * thread teams.
+ */
+template <typename ExeSpace = Kokkos::DefaultExecutionSpace>
+class TeamUtils
+{
+  int _team_size, _num_teams;
+
+public:
+  template <typename TeamPolicy>
+  TeamUtils(const TeamPolicy& policy) : _team_size(0)
+  {
+    const int max_threads = ExeSpace::concurrency();
+    const int team_size = policy.team_size();
+    _num_teams = max_threads / team_size;
+    _team_size = max_threads / _num_teams;
+  }
+
+  // How many thread teams can run concurrently
+  int get_num_concurrent_teams() const { return _num_teams; }
+
+  /*
+   * Of the C concurrently running teams, which "slot" is open
+   * for the given team.
+   */
+  template <typename MemberType>
+  KOKKOS_INLINE_FUNCTION
+  int get_workspace_idx(const MemberType& team_member) const
+  {
+#ifdef _OPENMP
```

Comment:
I was thinking about this. It's much better for this line to be `#ifdef KOKKOS_ENABLE_OPENMP`, so would you switch it to that? I put in the _OPENMP line to fix compile issues, but I see now that while _OPENMP is correct in `config_string`, it's not here; here, we're definitely dealing with explicitly Kokkos behavior. Thanks.